3.16: `@WithTestResource` starts all test resources (regression)

### Describe the bug

Since Quarkus 3.16 (including 3.16.1, 3.15.1 is fine) our integration test classes start literally test-resources from all literally all `@WithTestResource` annotations, even for those that are not (yet) executed.

Before 3.16(.1), only the test-resource for the currently running integration-test-class were started.

The major change in behavior that I could identify so far is in `io.quarkus.test.common.TestResourceManager#restrictToAnnotatedClass`, which now delegates to the `TestResourceClassEntryHandler` implementations (for `@QuarkusTestResource` and `@WithTestResource`). (Admittedly, [`restrictToAnnotatedClass` could never yield `true` before 3.16](https://github.com/quarkusio/quarkus/blob/a98a3f91fc06c959672b67ece75516bb59b994cd/test-framework/common/src/main/java/io/quarkus/test/common/TestResourceManager.java#L481).)

Since [restrictToAnnotatedClass() checks for `== RESTRICTED_TO_CLASS`](https://github.com/quarkusio/quarkus/blob/8a67e7919a7e373bd6edc6036c915aff23f25c1a/test-framework/common/src/main/java/io/quarkus/test/common/TestResourceManager.java#L494) on all annotations "everywhere", it initialized and starts all test resources. This can lead to e.g. 4 Keycloaks, 2 Minios and a bunch of other containers 🤷 

Our integration-tests are annotated using `@WithTestResource(value = xyz.class)`, leaving the default scope `MATCHING_RESOURCES` to leverage the `... Quarkus will not restart when running consecutive tests that use the same set of resources.` behavior.

The current workaround is to set the scope to `RESTRICTED_TO_CLASS`.

/cc @geoand 

### Expected behavior

* Only have the test-resources for the current test class running
* Group tests by test-resources, considering "initArgs"
* Don't start the "whole world" ;)


### Actual behavior

_No response_

### How to Reproduce?

I do have a reproducer, but it's still a (local) WIP to bump Quarkus to 3.16.1 in Nessie, including the workaround mentioned above.

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

Here's a reproducer: #44279
The added tests verify that only the expected test-resources are started (uses `TestInjector` for this), but fails because all test resources are started. 

There are three resources - all with scope `MATCHING_RESOURCES`.
One for "test one", one for "test two" and one that's shared but with different init-args.

Sample output:
```
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 2.301 s <<< FAILURE! -- in io.quarkus.it.extension.testresources.WithResourcesPoliciesFirstTest
[ERROR] io.quarkus.it.extension.testresources.WithResourcesPoliciesFirstTest.checkOnlyResource1started -- Time elapsed: 0.141 s <<< FAILURE!
org.opentest4j.AssertionFailedError: 

expected: ["resource1", null, "test-one"]
 but was: ["SomeResource1", "SomeResource2", "test-two"]
	at io.quarkus.it.extension.testresources.WithResourcesPoliciesFirstTest.checkOnlyResource1started(WithResourcesPoliciesFirstTest.java:28)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at io.quarkus.test.junit.QuarkusTestExtension.runExtensionMethod(QuarkusTestExtension.java:971)
	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestMethod(QuarkusTestExtension.java:821)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)

[INFO] Running io.quarkus.it.extension.testresources.WithResourcesPoliciesSecondTest
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.013 s <<< FAILURE! -- in io.quarkus.it.extension.testresources.WithResourcesPoliciesSecondTest
[ERROR] io.quarkus.it.extension.testresources.WithResourcesPoliciesSecondTest.checkOnlyResource1started -- Time elapsed: 0.001 s <<< FAILURE!
org.opentest4j.AssertionFailedError: 

expected: [null, "resource2", "test-two"]
 but was: ["SomeResource1", "SomeResource2", "test-two"]
	at io.quarkus.it.extension.testresources.WithResourcesPoliciesSecondTest.checkOnlyResource1started(WithResourcesPoliciesSecondTest.java:28)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at io.quarkus.test.junit.QuarkusTestExtension.runExtensionMethod(QuarkusTestExtension.java:971)
	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestMethod(QuarkusTestExtension.java:821)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
```

Maybe a bit clearer, when just running one of the added tests:
```
./mvnw -pl integration-tests/test-extension/tests install -Dtest=WithResourcesPoliciesFirstTest



...
SomeResource1 start
SharedResource start with arg 'test-one'
SomeResource2 start
SharedResource start with arg 'test-two'
...
[ERROR] io.quarkus.it.extension.testresources.WithResourcesPoliciesFirstTest.checkOnlyResource1started -- Time elapsed: 0.143 s <<< FAILURE!
org.opentest4j.AssertionFailedError: 

expected: ["resource1", null, "test-one"]
 but was: ["SomeResource1", "SomeResource2", "test-two"]
	at io.quarkus.it.extension.testresources.WithResourcesPoliciesFirstTest.checkOnlyResource1started(WithResourcesPoliciesFirstTest.java:28)

...

SomeResource1 stop
SharedResource stop
SomeResource2 stop
SharedResource stop
```

Starting `SomeResource2` and one of the `SharedResource` is wrong here.

--- a/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/testresources/SharedResource.java
+++ b/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/testresources/SharedResource.java
@@ -1,0 +1,44 @@
+package io.quarkus.it.extension.testresources;
+
+import static java.util.Objects.requireNonNull;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.util.Map;
+
+import io.quarkus.test.common.QuarkusTestResourceLifecycleManager;
+
+public class SharedResource implements QuarkusTestResourceLifecycleManager {
+    private String argument;
+
+    @Override
+    public void init(Map<String, String> initArgs) {
+        this.argument = requireNonNull(initArgs.get("resource.arg"));
+    }
+
+    @Override
+    public Map<String, String> start() {
+        System.err.println(getClass().getSimpleName() + " start with arg '" + argument + "'");
+        return Map.of();
+    }
+
+    @Override
+    public void stop() {
+        System.err.println(getClass().getSimpleName() + " stop");
+    }
+
+    @Override
+    public void inject(TestInjector testInjector) {
+        testInjector.injectIntoFields(argument,
+                new TestInjector.AnnotatedAndMatchesType(SharedResourceAnnotation.class, String.class));
+    }
+
+    @Documented
+    @Retention(RetentionPolicy.RUNTIME)
+    @Target(ElementType.FIELD)
+    public @interface SharedResourceAnnotation {
+    }
+}

--- a/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/testresources/SomeResource1.java
+++ b/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/testresources/SomeResource1.java
@@ -1,0 +1,35 @@
+package io.quarkus.it.extension.testresources;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.util.Map;
+
+import io.quarkus.test.common.QuarkusTestResourceLifecycleManager;
+
+public class SomeResource1 implements QuarkusTestResourceLifecycleManager {
+    @Override
+    public Map<String, String> start() {
+        System.err.println(getClass().getSimpleName() + " start");
+        return Map.of();
+    }
+
+    @Override
+    public void stop() {
+        System.err.println(getClass().getSimpleName() + " stop");
+    }
+
+    @Override
+    public void inject(TestInjector testInjector) {
+        testInjector.injectIntoFields(getClass().getSimpleName(),
+                new TestInjector.AnnotatedAndMatchesType(Resource1Annotation.class, String.class));
+    }
+
+    @Documented
+    @Retention(RetentionPolicy.RUNTIME)
+    @Target(ElementType.FIELD)
+    public @interface Resource1Annotation {
+    }
+}

--- a/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/testresources/SomeResource2.java
+++ b/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/testresources/SomeResource2.java
@@ -1,0 +1,35 @@
+package io.quarkus.it.extension.testresources;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.util.Map;
+
+import io.quarkus.test.common.QuarkusTestResourceLifecycleManager;
+
+public class SomeResource2 implements QuarkusTestResourceLifecycleManager {
+    @Override
+    public Map<String, String> start() {
+        System.err.println(getClass().getSimpleName() + " start");
+        return Map.of();
+    }
+
+    @Override
+    public void stop() {
+        System.err.println(getClass().getSimpleName() + " stop");
+    }
+
+    @Override
+    public void inject(TestInjector testInjector) {
+        testInjector.injectIntoFields(getClass().getSimpleName(),
+                new TestInjector.AnnotatedAndMatchesType(Resource2Annotation.class, String.class));
+    }
+
+    @Documented
+    @Retention(RetentionPolicy.RUNTIME)
+    @Target(ElementType.FIELD)
+    public @interface Resource2Annotation {
+    }
+}

--- a/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/testresources/WithResourcesPoliciesFirstTest.java
+++ b/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/testresources/WithResourcesPoliciesFirstTest.java
@@ -1,0 +1,31 @@
+package io.quarkus.it.extension.testresources;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.Arrays;
+
+import org.junit.jupiter.api.Test;
+
+import io.quarkus.test.common.ResourceArg;
+import io.quarkus.test.common.TestResourceScope;
+import io.quarkus.test.common.WithTestResource;
+import io.quarkus.test.junit.QuarkusTest;
+
+@QuarkusTest
+@WithTestResource(value = SomeResource1.class, scope = TestResourceScope.MATCHING_RESOURCES)
+@WithTestResource(value = SharedResource.class, scope = TestResourceScope.MATCHING_RESOURCES, initArgs = {
+        @ResourceArg(name = "resource.arg", value = "test-one") })
+public class WithResourcesPoliciesFirstTest {
+    @SomeResource1.Resource1Annotation
+    String resource1;
+    @SomeResource2.Resource2Annotation
+    String resource2;
+    @SharedResource.SharedResourceAnnotation
+    String sharedResource;
+
+    @Test
+    public void checkOnlyResource1started() {
+        assertThat(Arrays.asList(resource1, resource2, sharedResource)).isEqualTo(
+                Arrays.asList("SomeResource1", null, "test-one"));
+    }
+}

--- a/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/testresources/WithResourcesPoliciesSecondTest.java
+++ b/integration-tests/test-extension/tests/src/test/java/io/quarkus/it/extension/testresources/WithResourcesPoliciesSecondTest.java
@@ -1,0 +1,31 @@
+package io.quarkus.it.extension.testresources;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.Arrays;
+
+import org.junit.jupiter.api.Test;
+
+import io.quarkus.test.common.ResourceArg;
+import io.quarkus.test.common.TestResourceScope;
+import io.quarkus.test.common.WithTestResource;
+import io.quarkus.test.junit.QuarkusTest;
+
+@QuarkusTest
+@WithTestResource(value = SomeResource2.class, scope = TestResourceScope.MATCHING_RESOURCES)
+@WithTestResource(value = SharedResource.class, scope = TestResourceScope.MATCHING_RESOURCES, initArgs = {
+        @ResourceArg(name = "resource.arg", value = "test-two") })
+public class WithResourcesPoliciesSecondTest {
+    @SomeResource1.Resource1Annotation
+    String resource1;
+    @SomeResource2.Resource2Annotation
+    String resource2;
+    @SharedResource.SharedResourceAnnotation
+    String sharedResource;
+
+    @Test
+    public void checkOnlyResource1started() {
+        assertThat(Arrays.asList(resource1, resource2, sharedResource)).isEqualTo(
+                Arrays.asList(null, "SomeResource2", "test-two"));
+    }
+}

--- a/test-framework/common/src/main/java/io/quarkus/test/common/TestResourceManager.java
+++ b/test-framework/common/src/main/java/io/quarkus/test/common/TestResourceManager.java
@@ -439,15 +439,15 @@ public class TestResourceManager implements Closeable {
 
     private static Collection<AnnotationInstance> findTestResourceInstancesOfClass(Class<?> testClass, IndexView index) {
         // collect all test supertypes for matching per-test targets
-        Set<String> testClasses = new HashSet<>();
+        Set<String> currentTestClassHierarchy = new HashSet<>();
         Class<?> current = testClass;
         while (current != Object.class) {
-            testClasses.add(current.getName());
+            currentTestClassHierarchy.add(current.getName());
             current = current.getSuperclass();
         }
         current = testClass.getEnclosingClass();
         while (current != null) {
-            testClasses.add(current.getName());
+            currentTestClassHierarchy.add(current.getName());
             current = current.getEnclosingClass();
         }
 
@@ -455,7 +455,7 @@ public class TestResourceManager implements Closeable {
 
         for (DotName testResourceClasses : List.of(WITH_TEST_RESOURCE, QUARKUS_TEST_RESOURCE)) {
             for (AnnotationInstance annotation : index.getAnnotations(testResourceClasses)) {
-                if (keepTestResourceAnnotation(annotation, annotation.target().asClass(), testClasses)) {
+                if (keepTestResourceAnnotation(annotation, annotation.target().asClass(), currentTestClassHierarchy)) {
                     testResourceAnnotations.add(annotation);
                 }
             }
@@ -466,7 +466,8 @@ public class TestResourceManager implements Closeable {
             for (AnnotationInstance annotation : index.getAnnotations(testResourceListClasses)) {
                 for (AnnotationInstance nestedAnnotation : annotation.value().asNestedArray()) {
                     // keep the list target
-                    if (keepTestResourceAnnotation(nestedAnnotation, annotation.target().asClass(), testClasses)) {
+                    if (keepTestResourceAnnotation(nestedAnnotation, annotation.target().asClass(),
+                            currentTestClassHierarchy)) {
                         testResourceAnnotations.add(nestedAnnotation);
                     }
                 }
@@ -477,21 +478,22 @@ public class TestResourceManager implements Closeable {
     }
 
     private static boolean keepTestResourceAnnotation(AnnotationInstance annotation, ClassInfo targetClass,
-            Set<String> testClasses) {
+            Set<String> currentTestClassHierarchy) {
         if (targetClass.isAnnotation()) {
             // meta-annotations have already been handled in collectMetaAnnotations
             return false;
         }
 
         if (restrictToAnnotatedClass(annotation)) {
-            return testClasses.contains(targetClass.name().toString('.'));
+            return currentTestClassHierarchy.contains(targetClass.name().toString('.'));
         }
 
         return true;
     }
 
     private static boolean restrictToAnnotatedClass(AnnotationInstance annotation) {
-        return TestResourceClassEntryHandler.determineScope(annotation) == RESTRICTED_TO_CLASS;
+        return TestResourceClassEntryHandler.determineScope(annotation) == RESTRICTED_TO_CLASS
+                || TestResourceClassEntryHandler.determineScope(annotation) == MATCHING_RESOURCES;
     }
 
     /**

--- a/test-framework/common/src/main/java/io/quarkus/test/common/TestResourceManager.java
+++ b/test-framework/common/src/main/java/io/quarkus/test/common/TestResourceManager.java
@@ -108,7 +108,7 @@ public class TestResourceManager implements Closeable {
         this.testResourceComparisonInfo = new HashSet<>();
         for (TestResourceClassEntry uniqueEntry : uniqueEntries) {
             testResourceComparisonInfo.add(new TestResourceComparisonInfo(
-                    uniqueEntry.testResourceLifecycleManagerClass().getName(), uniqueEntry.getScope()));
+                    uniqueEntry.testResourceLifecycleManagerClass().getName(), uniqueEntry.getScope(), uniqueEntry.args));
         }
 
         Set<TestResourceClassEntry> remainingUniqueEntries = initParallelTestResources(uniqueEntries);
@@ -326,7 +326,8 @@ public class TestResourceManager implements Closeable {
         }
         Set<TestResourceManager.TestResourceComparisonInfo> result = new HashSet<>(uniqueEntries.size());
         for (TestResourceClassEntry entry : uniqueEntries) {
-            result.add(new TestResourceComparisonInfo(entry.testResourceLifecycleManagerClass().getName(), entry.getScope()));
+            result.add(new TestResourceComparisonInfo(entry.testResourceLifecycleManagerClass().getName(), entry.getScope(),
+                    entry.args));
         }
         return result;
     }
@@ -518,7 +519,7 @@ public class TestResourceManager implements Closeable {
             return false;
         }
 
-        if (hasRestrictedToClassScope(existing) || hasRestrictedToClassScope(next)) {
+        if (anyResourceRestrictedToClass(existing) || anyResourceRestrictedToClass(next)) {
             return true;
         }
 
@@ -540,8 +541,8 @@ public class TestResourceManager implements Closeable {
         return false;
     }
 
-    private static boolean hasRestrictedToClassScope(Set<TestResourceComparisonInfo> existing) {
-        for (TestResourceComparisonInfo info : existing) {
+    private static boolean anyResourceRestrictedToClass(Set<TestResourceComparisonInfo> testResources) {
+        for (TestResourceComparisonInfo info : testResources) {
             if (info.scope == RESTRICTED_TO_CLASS) {
                 return true;
             }
@@ -605,7 +606,8 @@ public class TestResourceManager implements Closeable {
         }
     }
 
-    public record TestResourceComparisonInfo(String testResourceLifecycleManagerClass, TestResourceScope scope) {
+    public record TestResourceComparisonInfo(String testResourceLifecycleManagerClass, TestResourceScope scope,
+            Map<String, String> args) {
 
     }
 

--- a/test-framework/common/src/test/java/io/quarkus/test/common/TestResourceManagerReloadTest.java
+++ b/test-framework/common/src/test/java/io/quarkus/test/common/TestResourceManagerReloadTest.java
@@ -6,6 +6,7 @@ import static org.junit.jupiter.api.Assertions.assertFalse;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
 import java.util.Collections;
+import java.util.Map;
 import java.util.Set;
 
 import org.junit.jupiter.api.Test;
@@ -22,66 +23,87 @@ public class TestResourceManagerReloadTest {
     @Test
     public void differentCount() {
         assertTrue(testResourcesRequireReload(Collections.emptySet(),
-                Set.of(new TestResourceComparisonInfo("test", RESTRICTED_TO_CLASS))));
+                Set.of(new TestResourceComparisonInfo("test", RESTRICTED_TO_CLASS, Map.of()))));
 
-        assertTrue(testResourcesRequireReload(Set.of(new TestResourceComparisonInfo("test", RESTRICTED_TO_CLASS)),
+        assertTrue(testResourcesRequireReload(Set.of(new TestResourceComparisonInfo("test", RESTRICTED_TO_CLASS, Map.of())),
                 Collections.emptySet()));
     }
 
     @Test
     public void sameSingleRestrictedToClassResource() {
         assertTrue(testResourcesRequireReload(
-                Set.of(new TestResourceComparisonInfo("test", RESTRICTED_TO_CLASS)),
-                Set.of(new TestResourceComparisonInfo("test", RESTRICTED_TO_CLASS))));
+                Set.of(new TestResourceComparisonInfo("test", RESTRICTED_TO_CLASS, Map.of())),
+                Set.of(new TestResourceComparisonInfo("test", RESTRICTED_TO_CLASS, Map.of()))));
     }
 
     @Test
     public void sameSingleMatchingResource() {
         assertFalse(testResourcesRequireReload(
-                Set.of(new TestResourceComparisonInfo("test", MATCHING_RESOURCES)),
-                Set.of(new TestResourceComparisonInfo("test", MATCHING_RESOURCES))));
+                Set.of(new TestResourceComparisonInfo("test", MATCHING_RESOURCES, Map.of())),
+                Set.of(new TestResourceComparisonInfo("test", MATCHING_RESOURCES, Map.of()))));
+    }
+
+    @Test
+    public void sameSingleMatchingResourceWithArgs() {
+        assertFalse(testResourcesRequireReload(
+                Set.of(new TestResourceComparisonInfo("test", MATCHING_RESOURCES, Map.of("a", "b"))),
+                Set.of(new TestResourceComparisonInfo("test", MATCHING_RESOURCES, Map.of("a", "b")))));
+    }
+
+    @Test
+    public void sameSingleResourceDifferentArgs() {
+        assertTrue(testResourcesRequireReload(
+                Set.of(new TestResourceComparisonInfo("test", MATCHING_RESOURCES, Map.of())),
+                Set.of(new TestResourceComparisonInfo("test", MATCHING_RESOURCES, Map.of("a", "b")))));
+    }
+
+    @Test
+    public void sameSingleResourceDifferentArgValues() {
+        assertTrue(testResourcesRequireReload(
+                Set.of(new TestResourceComparisonInfo("test", MATCHING_RESOURCES, Map.of("a", "b"))),
+                Set.of(new TestResourceComparisonInfo("test", MATCHING_RESOURCES, Map.of("a", "c")))));
     }
 
     @Test
     public void differentSingleMatchingResource() {
         assertTrue(testResourcesRequireReload(
-                Set.of(new TestResourceComparisonInfo("test", MATCHING_RESOURCES)),
-                Set.of(new TestResourceComparisonInfo("test2", MATCHING_RESOURCES))));
+                Set.of(new TestResourceComparisonInfo("test", MATCHING_RESOURCES, Map.of())),
+                Set.of(new TestResourceComparisonInfo("test2", MATCHING_RESOURCES, Map.of()))));
     }
 
     @Test
     public void sameMultipleMatchingResource() {
         assertFalse(testResourcesRequireReload(
                 Set.of(
-                        new TestResourceComparisonInfo("test", MATCHING_RESOURCES),
-                        new TestResourceComparisonInfo("test2", MATCHING_RESOURCES),
-                        new TestResourceComparisonInfo("test3", GLOBAL)),
-                Set.of(new TestResourceComparisonInfo("test3", GLOBAL),
-                        new TestResourceComparisonInfo("test2", MATCHING_RESOURCES),
-                        new TestResourceComparisonInfo("test", MATCHING_RESOURCES))));
+                        new TestResourceComparisonInfo("test", MATCHING_RESOURCES, Map.of()),
+                        new TestResourceComparisonInfo("test2", MATCHING_RESOURCES, Map.of()),
+                        new TestResourceComparisonInfo("test3", GLOBAL, Map.of())),
+                Set.of(new TestResourceComparisonInfo("test3", GLOBAL, Map.of()),
+                        new TestResourceComparisonInfo("test2", MATCHING_RESOURCES, Map.of()),
+                        new TestResourceComparisonInfo("test", MATCHING_RESOURCES, Map.of()))));
     }
 
     @Test
     public void differentMultipleMatchingResource() {
         assertTrue(testResourcesRequireReload(
                 Set.of(
-                        new TestResourceComparisonInfo("test", MATCHING_RESOURCES),
-                        new TestResourceComparisonInfo("test2", MATCHING_RESOURCES),
-                        new TestResourceComparisonInfo("test3", GLOBAL)),
-                Set.of(new TestResourceComparisonInfo("test3", GLOBAL),
-                        new TestResourceComparisonInfo("test2", MATCHING_RESOURCES),
-                        new TestResourceComparisonInfo("TEST", MATCHING_RESOURCES))));
+                        new TestResourceComparisonInfo("test", MATCHING_RESOURCES, Map.of()),
+                        new TestResourceComparisonInfo("test2", MATCHING_RESOURCES, Map.of()),
+                        new TestResourceComparisonInfo("test3", GLOBAL, Map.of())),
+                Set.of(new TestResourceComparisonInfo("test3", GLOBAL, Map.of()),
+                        new TestResourceComparisonInfo("test2", MATCHING_RESOURCES, Map.of()),
+                        new TestResourceComparisonInfo("TEST", MATCHING_RESOURCES, Map.of()))));
     }
 
     @Test
     public void differentGlobalMultipleMatchingResource() {
         assertTrue(testResourcesRequireReload(
                 Set.of(
-                        new TestResourceComparisonInfo("test", MATCHING_RESOURCES),
-                        new TestResourceComparisonInfo("test2", MATCHING_RESOURCES),
-                        new TestResourceComparisonInfo("test4", GLOBAL)),
-                Set.of(new TestResourceComparisonInfo("test3", GLOBAL),
-                        new TestResourceComparisonInfo("test2", MATCHING_RESOURCES),
-                        new TestResourceComparisonInfo("TEST", MATCHING_RESOURCES))));
+                        new TestResourceComparisonInfo("test", MATCHING_RESOURCES, Map.of()),
+                        new TestResourceComparisonInfo("test2", MATCHING_RESOURCES, Map.of()),
+                        new TestResourceComparisonInfo("test4", GLOBAL, Map.of())),
+                Set.of(new TestResourceComparisonInfo("test3", GLOBAL, Map.of()),
+                        new TestResourceComparisonInfo("test2", MATCHING_RESOURCES, Map.of()),
+                        new TestResourceComparisonInfo("TEST", MATCHING_RESOURCES, Map.of()))));
     }
 }

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/AbstractTestWithCallbacksExtension.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/AbstractTestWithCallbacksExtension.java
@@ -100,7 +100,7 @@ public abstract class AbstractTestWithCallbacksExtension {
         invokeCallbacks(afterAllCallbacks, "afterAll", clazz, testContext);
     }
 
-    protected void populateCallbacks(ClassLoader classLoader) throws ClassNotFoundException {
+    protected static void clearCallbacks() {
         beforeClassCallbacks = new ArrayList<>();
         afterConstructCallbacks = new ArrayList<>();
         beforeEachCallbacks = new ArrayList<>();
@@ -108,6 +108,10 @@ public abstract class AbstractTestWithCallbacksExtension {
         afterTestCallbacks = new ArrayList<>();
         afterEachCallbacks = new ArrayList<>();
         afterAllCallbacks = new ArrayList<>();
+    }
+
+    protected void populateCallbacks(ClassLoader classLoader) throws ClassNotFoundException {
+        clearCallbacks();
 
         ServiceLoader<?> quarkusTestBeforeClassLoader = ServiceLoader
                 .load(Class.forName(QuarkusTestBeforeClassCallback.class.getName(), false, classLoader), classLoader);

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/IntegrationTestExtensionState.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/IntegrationTestExtensionState.java
@@ -8,11 +8,11 @@ import io.quarkus.test.common.TestResourceManager;
 
 public class IntegrationTestExtensionState extends QuarkusTestExtensionState {
 
-    private Map<String, String> sysPropRestore;
+    private final Map<String, String> sysPropRestore;
 
     public IntegrationTestExtensionState(TestResourceManager testResourceManager, Closeable resource,
-            Map<String, String> sysPropRestore) {
-        super(testResourceManager, resource);
+            Runnable clearCallbacks, Map<String, String> sysPropRestore) {
+        super(testResourceManager, resource, clearCallbacks);
         this.sysPropRestore = sysPropRestore;
     }
 

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusIntegrationTestExtension.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusIntegrationTestExtension.java
@@ -13,12 +13,10 @@ import static io.quarkus.test.junit.IntegrationTestUtil.getSysPropsToRestore;
 import static io.quarkus.test.junit.IntegrationTestUtil.handleDevServices;
 import static io.quarkus.test.junit.IntegrationTestUtil.readQuarkusArtifactProperties;
 import static io.quarkus.test.junit.IntegrationTestUtil.startLauncher;
-import static io.quarkus.test.junit.TestResourceUtil.testResourcesRequireReload;
 import static io.quarkus.test.junit.TestResourceUtil.TestResourceManagerReflections.copyEntriesFromProfile;
 
 import java.io.Closeable;
 import java.io.File;
-import java.io.IOException;
 import java.lang.reflect.Field;
 import java.nio.file.Path;
 import java.time.Duration;
@@ -108,7 +106,6 @@ public class QuarkusIntegrationTestExtension extends AbstractQuarkusTestWithCont
 
         } else {
             throwBootFailureException();
-            return;
         }
     }
 
@@ -305,7 +302,7 @@ public class QuarkusIntegrationTestExtension extends AbstractQuarkusTestWithCont
             Closeable resource = new IntegrationTestExtensionStateResource(launcher,
                     devServicesLaunchResult.getCuratedApplication());
             IntegrationTestExtensionState state = new IntegrationTestExtensionState(testResourceManager, resource,
-                    sysPropRestore);
+                    AbstractTestWithCallbacksExtension::clearCallbacks, sysPropRestore);
             testHttpEndpointProviders = TestHttpEndpointProvider.load();
 
             return state;
@@ -467,7 +464,7 @@ public class QuarkusIntegrationTestExtension extends AbstractQuarkusTestWithCont
         }
 
         @Override
-        public void close() throws IOException {
+        public void close() {
             if (launcher != null) {
                 try {
                     launcher.close();

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java
@@ -3,7 +3,6 @@ package io.quarkus.test.junit;
 import static io.quarkus.test.junit.IntegrationTestUtil.activateLogging;
 
 import java.io.Closeable;
-import java.io.IOException;
 import java.lang.management.ManagementFactory;
 import java.lang.management.ThreadInfo;
 import java.lang.reflect.Constructor;
@@ -272,7 +271,7 @@ public class QuarkusTestExtension extends AbstractJvmQuarkusTestExtension
 
             Closeable shutdownTask = new Closeable() {
                 @Override
-                public void close() throws IOException {
+                public void close() {
                     TracingHandler.quarkusStopping();
                     try {
                         runningQuarkusApplication.close();
@@ -295,8 +294,7 @@ public class QuarkusTestExtension extends AbstractJvmQuarkusTestExtension
                     }
                 }
             };
-            ExtensionState state = new ExtensionState(testResourceManager, shutdownTask);
-            return state;
+            return new ExtensionState(testResourceManager, shutdownTask, AbstractTestWithCallbacksExtension::clearCallbacks);
         } catch (Throwable e) {
             if (!InitialConfigurator.DELAYED_HANDLER.isActivated()) {
                 activateLogging();
@@ -1176,12 +1174,12 @@ public class QuarkusTestExtension extends AbstractJvmQuarkusTestExtension
 
     public static class ExtensionState extends QuarkusTestExtensionState {
 
-        public ExtensionState(Closeable testResourceManager, Closeable resource) {
-            super(testResourceManager, resource);
+        public ExtensionState(Closeable testResourceManager, Closeable resource, Runnable clearCallbacks) {
+            super(testResourceManager, resource, clearCallbacks);
         }
 
         @Override
-        protected void doClose() throws IOException {
+        protected void doClose() {
             ClassLoader old = Thread.currentThread().getContextClassLoader();
             if (runningQuarkusApplication != null) {
                 Thread.currentThread().setContextClassLoader(runningQuarkusApplication.getClassLoader());

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtensionState.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtensionState.java
@@ -15,11 +15,13 @@ public class QuarkusTestExtensionState implements ExtensionContext.Store.Closeab
     protected final Closeable testResourceManager;
     protected final Closeable resource;
     private final Thread shutdownHook;
+    private final Runnable clearCallbacks;
     private Throwable testErrorCause;
 
-    public QuarkusTestExtensionState(Closeable testResourceManager, Closeable resource) {
+    public QuarkusTestExtensionState(Closeable testResourceManager, Closeable resource, Runnable clearCallbacks) {
         this.testResourceManager = testResourceManager;
         this.resource = resource;
+        this.clearCallbacks = clearCallbacks;
         this.shutdownHook = new Thread(new Runnable() {
             @Override
             public void run() {
@@ -40,6 +42,7 @@ public class QuarkusTestExtensionState implements ExtensionContext.Store.Closeab
     public void close() throws IOException {
         if (closed.compareAndSet(false, true)) {
             doClose();
+            clearCallbacks.run();
 
             try {
                 Runtime.getRuntime().removeShutdownHook(shutdownHook);

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/TestResourceUtil.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/TestResourceUtil.java
@@ -192,8 +192,10 @@ final class TestResourceUtil {
                     if (originalTestResourceScope != null) {
                         testResourceScope = TestResourceScope.valueOf(originalTestResourceScope.toString());
                     }
+                    Object originalArgs = entry.getClass().getMethod("args").invoke(entry);
+                    Map<String, String> args = (Map<String, String>) originalArgs;
                     result.add(new TestResourceManager.TestResourceComparisonInfo(testResourceLifecycleManagerClass,
-                            testResourceScope));
+                            testResourceScope, args));
                 }
 
                 return result;

